### PR TITLE
INVALID* (0.8822 BPB mean) Medusa: Unstable S2 — DeltaNet Crawler, Legal  10mb. .77bpb single seed.

### PR DESCRIPTION
**Pull request by newjordan**
medusas_gaze_compiled_v1

## Summary

UNSTABLE PROBLEM SOLVED - DELTANET needs re-work as is. 

With DeltaNEt OFF This submission scored a val_bpb:1.18234857 @ 10mb. Will try deltanet adaptions in a different manner, as this one was "learning to cheat better" every cycle (i think)

Legal resubmission of PR #1028 (Medusa: Unstable, 0.9984 BPB mean), which was flagged because GPTQ calibration read training data after the 600s wallclock cap.

**Fix:** `GPTQ_RESERVE_MS=30000` stops the training loop at ~570s, so GPTQ calibration (~12s) completes within the 600s budget. Log confirms timing:
```
stopping_early: wallclock_cap train_time:570093ms step:4628/20000
gptq:loop-aware calibrated 41 layers in 11.5s
```

All hyperparameters identical to PR #1028 / Medusa_IV.

## Results

| Seed | SW BPB | Post-EMA BPB | Int6 Roundtrip |
|------|-------:|-------------:|---------------:|
| 300  | 1.0251 | 0.6484 | 0.8987 |
| 444  | 0.8469 | 0.4330 | 0.7159 |
| 4    | **0.7744** | 0.4339 | 0.6271 |
| **Mean** | **0.8822** | | |
| **Std dev** | ~0.105 | | |

3-seed mean improved from 0.9984 → 0.8822 vs the flagged submission.

Look at the EMA unravel, and the inconsistency! Its a puzzle waiting to be solved and the answer could be an invalid system. 

## Architecture

- 4 flat layers + 1 crawler × 4 loops (Frugendorff), INST_DIM=32
- DELTA_NET_HEADS=4, `chunk_delta_rule` (fla.ops.delta_rule)
- Loop-aware 2-phase GPTQ (41 layers), int6+zstd
- EMA_START_STEP=4400, EMA_DECAY=0.99
- ~9.8MB artifact, 8xH100 SXM, 600s training

## Code

`train_gpt.py` trimmed: ~632 lines of dead n-gram evaluation code removed (NGRAM_EVAL_ORDER=0 in this config). No logic changes.

## Known Issues

High cross-seed variance (std dev ~0.105) from DeltaNet heads. Root causes identified (state dtype bug, quantization unravel through 4 crawler loops). Stabilization is active research.

Delta net black box information could need better integration.

### CODE REVIEW

**Comment by Eppie** [comment]
It took me quite a few back-and-forths with Opus, who initially was convinced this was a legal/valid implementation, but I think I found a problem here. I'll let Opus explain:

---
The crawler block uses causal attention, so at the end of loop 1, position `t`'s hidden state only knows about tokens `0..t`. That's fine.

But the DeltaNet writes every position's hidden state into the shared matrix `S`, sequentially, across the full sequence. After loop 1 finishes, `S` contains information from all `T` positions.

Loop 2 starts with that same `S`. When position `t` reads from `S` in loop 2, it can retrieve information that was written by positions `t+1`, `t+2`, ..., `T-1` during loop 1.

Position `t` is now making its prediction with knowledge of future tokens.

That's the violation. The DeltaNet state acts as a backdoor that bypasses the causal mask.

**Comment by newjordan** [comment]
> It took me quite a few back-and-forths with Opus, who initially was convinced this was a legal/valid implementation, but I think I found a problem here. I'll let Opus explain:
> 
> 
> 
> ---
> 
> The crawler block uses causal attention, so at the end of loop 1, position `t`'s hidden state only knows about tokens `0..t`. That's fine.
> 
> 
> 
> But the DeltaNet writes every position's hidden state into the shared matrix `S`, sequentially, across the full sequence. After loop 1 finishes, `S` contains information from all `T` positions.
> 
> 
> 
> Loop 2 starts with that same `S`. When position `t` reads from `S` in loop 2, it can retrieve information that was written by positions `t+1`, `t+2`, ..., `T-1` during loop 1.
> 
> 
> 
> Position `t` is now making its prediction with knowledge of future tokens.
> 
> 
> 
> That's the violation. The DeltaNet state acts as a backdoor that bypasses the causal mask.

Maybe, it's a triton node and that's kinda how they work, they are a black box that pass data. Question is if that data is forward looking, in my opinion, it's not. Even if it was, that's a small adjustment to a variable and the system still needs to be tested etc. the basic architecture works outside of some small variable in the triton node...

**Comment by MatoTeziTanka** [comment]
### Community Review — INVALID* (0.8822 BPB mean) Medusa: Unstable S2 — DeltaNet Crawler, Legal  10mb. .77bpb single seed.

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## PR #1047 — Medusa: Unstable S2 (Legal Resubmission of #1028) **Head SHA:** 99d34869c2ced0dd13b353f075765cac14246961 **Summary:** Legal resubmission fixing the GPTQ timing violation from PR #1028. Training stops 30s early (`GPTQ_RESERVE_MS=30000`), GPTQ calibration completes at ~582s within the 600s wallclock budget. --- ### Check 1: ILLEGAL n-gram family bug — CLEAR `BigramHashEmbedding.bigram_hash()` (lines 728–734): hashes `XOR(36313 * t[i], 27191 * t[i-1])` using only **context tokens** (prior token). No target token (`y`) is XOR'd into the lookup key. This is a learned embedding applied to *inputs*, not a count-table hash that encodes the target. The `TrainNgramOracle` / `TrainNgramOracleGPU` classes (lines 823–992) use `full_key = ctx_hash XOR (tgt * prime)` to count training corpus n-gram occurrences, which is the standard BigramHash structure — target is XOR'd into the **full_key for counting in training**, not into the input embedding. **NGRAM_EVAL_ORDER=0** in run.sh (line ~100 of run.sh), so eval-time n-gram interpolation is OFF. **No ILLEGAL n-gram family bug.** ### Check 2: ILLEGAL Pre-Quant TTT — CLEAR No test-time training loop over `val_tokens`. The `val_tokens` tensor is used only in `eval_val()` (lines 352–397) and `eval_val_sliding()` (lines 1854–1913) under `torch.inference_mode()` with `model.eval()` — no gradient computation, no optimizer steps. GPTQ calibration (`gptq_calibrate`, lines 1993–2025) reads from `train_pattern` (training data), not `val_tokens`. **No Pre-Quant TTT violation.** ### Check 3: LEGAL score-first TTT — N/A No TTT mechanism present at all. ### Check 4: SCORED REGION SLOT — N/A No scored-region reservation pattern present. ### Assessment Pure neural model: DeltaNet crawler architecture with BigramHash embedding, loop-aware GPTQ quantization, EMA....

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
